styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -736,25 +736,48 @@ void postResource(Resource resource) throws IOException {
     storeWithRetries(() -> runInTransaction(transaction -> {
       transaction.store(resource);
       return null;
-      // TODO store just in one place, eliminate one of the two calls ^?
     }));
   }
 
   List<Resource> getResources() throws IOException {
-    final EntityQuery query = Query.newEntityQueryBuilder().setKind(KIND_RESOURCE).build();
+    final EntityQuery query = Query.newEntityQueryBuilder().setKind(KIND_COUNTER_LIMIT).build();
     final List<Resource> resources = Lists.newArrayList();
     datastore.query(query, entity ->
         resources.add(entityToResource(entity)));
     return resources;
   }
 
-  private Resource entityToResource(Entity entity) {
-    return Resource.create(entity.getKey().getName(), entity.getLong(PROPERTY_CONCURRENCY));
+  /**
+   * Delete resource by id. Deletes both counter shards and counter limit if it exists.
+   *
+   * <p>Due to Datastore limitations (modify max 25 entity groups per transaction),
+   * deletion of shards is done in batches of 25 shards.
+   */
+  void deleteResource(String id) throws IOException {
+    storeWithRetries(() -> runInTransaction(tx -> {
+      tx.deleteCounterLimit(id);
+      return null;
+    }));
+    deleteShardsForCounter(id);
   }
 
-  void deleteResource(String id) throws IOException {
-    final Key key = datastore.newKeyFactory().setKind(KIND_RESOURCE).newKey(id);
-    datastore.delete(key);
+  private void deleteShardsForCounter(String counterId) throws IOException {
+    final List<Key> shards = new ArrayList<>();
+    datastore.query(EntityQuery.newEntityQueryBuilder()
+        .setKind(KIND_COUNTER_SHARD)
+        .setFilter(PropertyFilter.eq(PROPERTY_COUNTER_ID, counterId))
+        .build(), entity -> shards.add(entity.getKey()));
+
+    for (List<Key> batch : Lists.partition(shards, MAX_NUMBER_OF_ENTITY_GROUPS_IN_ONE_TRANSACTION)) {
+      storeWithRetries(() -> runInTransaction(transaction -> {
+        datastore.delete(batch.toArray(new Key[0]));
```

Comment:
This doesn't actually perform the delete in a transaction as it is calling `datastore.delete()` instead of `transaction.delete()` but in this case I'm not sure it is even necessary to use transactions, seeing as we anyway cannot atomically delete all of the shards atomically in one transaction.

Reply:
It feels somewhat more reliable using transaction than plain batch deletion.